File: cpp/linkedLists/src/linkedLists.cpp
#include "linkedLists.h"

SinglyLinkedList::Node::Node(){
    next = nullptr;
    value = nullptr;
}

SinglyLinkedList::Node::Node(float init_value){
    next = nullptr;
    value = new float(init_value);
}

SinglyLinkedList::Node::~Node(){
    delete value;
}

SinglyLinkedList::SinglyLinkedList(){
    head = nullptr;
}

SinglyLinkedList::SinglyLinkedList(std::vector<float> vec){
    head = nullptr;
    std::reverse(vec.begin(), vec.end());
    for(int i = 0; i < vec.size(); i++){
        this->pushHead(vec[i]);
    }
}

SinglyLinkedList::~SinglyLinkedList(){
    Node* current = head;
    while (current != nullptr){
        Node* prev_node  = current;
        current = current->next;
        delete prev_node;
    }
}

void SinglyLinkedList::pushHead(float value){
    Node* node = new Node(value);
    if (head == nullptr){
        head = node;
    } else {
        node->next = head;
        head = node;
    }
}

float SinglyLinkedList::popHead(){
    if (head == nullptr){
        std::cerr << "popHead: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    }
    Node* old_head = head;
    if (head->next == nullptr) {
        head = nullptr;
    } else {
        head = head->next;
    }

    float value = *(old_head->value);
    delete old_head;

    return value;
}

float SinglyLinkedList::getHead(){
    if (head == nullptr){
        std::cerr << "getHead: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    } else {
        float value = *(head->value);
        return value;
    }
}

float SinglyLinkedList::getTail(){
    if (head == nullptr){
        std::cerr << "getTail: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    } else {
        Node* tail = this->getTailNode();
        float value = *(tail->value);
        return value;
    }
}

int SinglyLinkedList::getLength(){
    Node* current = head;
    int n = 0;
    while (current != nullptr){
        current = current->next;
        n++;
    }
    return n;
}

void SinglyLinkedList::pushTail(float value){
    Node* node = new Node(value);
    if (head == nullptr){
        head = node;
    } else {
        Node* tail = this->getTailNode();
        tail->next = node;
    }
}

float SinglyLinkedList::popTail(){
    if (head == nullptr){
        std::cerr << "popTail: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    } else if (head->next == nullptr) {
        float value = *(head->value);
        delete(head);
        head = nullptr;
        return value;
    } else {
        Node* current = head;
        Node* prev = nullptr;
        while (current->next != nullptr){
            prev = current;
            current = current->next;
        }
        prev->next = nullptr;
        float value = *(current->value);
        delete current;
        return value;
    }
}

SinglyLinkedList::Node* SinglyLinkedList::getTailNode(){
    Node* current = head;
    while (current->next != nullptr){
        current = current->next;
    }
    return current;
}

std::vector<float> SinglyLinkedList::getVector(){
    std::vector<float> vec;
    if (head == nullptr){
        return vec;
    } else {
        Node* current = head;
        while (current->next != nullptr) {
            vec.push_back(*(current->value));
            current = current->next;
        }
        vec.push_back(*(current->value));
        return vec;
    }
}

float SinglyLinkedList::valueAt(int index){
    int length = this->boundCheck(index);
    Node* current = head;
    for(int i = 0; i < index; i++){
        current = current->next;
    }
    float value = *(current->value);
    return value;
}

void SinglyLinkedList::deleteAt(int index){
    int length = this->boundCheck(index);
    if (index == 0){
        Node* new_head = head->next;
        delete head;
        head = new_head;
    } else {
        Node* current = head;
        for(int i = 0; i < index-1; i++){
            current = current->next;
        }
        Node* node_to_delete = current->next;
        if (node_to_delete->next != nullptr){
            Node* node_after_deletion = node_to_delete->next;
            current->next = node_after_deletion;
        } else {
            current->next = nullptr;
        }
        delete node_to_delete;
    }
}

void SinglyLinkedList::insertAt(int index, float value){
    int length = this->boundCheck(index);
    Node* new_node = new Node(value);
    if (index == 0){
        new_node->next = head;
        head = new_node;
    } else {
        Node* current = head;
        for(int i = 0; i < index-1; i++){
            current = current->next;
        }
        Node* node_after = current->next;
        current->next = new_node;
        new_node->next = node_after;
    }
}

int SinglyLinkedList::boundCheck(int index){
    int length = this->getLength();
    if (index > length - 1 or index < 0) {
        std::cerr << "boundCheck: Index " << index << " out of range for list of length " << length << std::endl;
        exit(EXIT_FAILURE);
    }
    return length;
}

void SinglyLinkedList::reverse(){
    int length = this->getLength();
    if (length == 0 or length == 1){
        return;
    }
    Node* node1 = head;
    Node* node2 = head->next;
    head->next = nullptr;
    Node* node3;
    while(node2->next != nullptr){
        node3 = node2->next;
        node2->next = node1;
        node1 = node2;
        node2 = node3;
    }
    node2->next = node1;
    head = node2;
}

////////////////////////////////////////////////////////////////

DoublyLinkedList::Node::Node(){
    next = nullptr;
    prev = nullptr;
    value = nullptr;
}

DoublyLinkedList::Node::Node(float init_value){
    next = nullptr;
    prev = nullptr;
    value = new float(init_value);
}

DoublyLinkedList::Node::~Node(){
    delete value;
}

DoublyLinkedList::DoublyLinkedList(){
    head = nullptr;
    tail = nullptr;
}

DoublyLinkedList::DoublyLinkedList(std::vector<float> vec){
    head = nullptr;
    tail = nullptr;
    std::reverse(vec.begin(), vec.end());
    for(int i = 0; i < vec.size(); i++){
        this->pushHead(vec[i]);
    }
}

DoublyLinkedList::~DoublyLinkedList(){
    Node* current = head;
    while (current != nullptr){
        Node* prev_node  = current;
        current = current->next;
        delete prev_node;
    }
}

void DoublyLinkedList::pushHead(float value){
    Node* node = new Node(value);
    if (head == nullptr){
        head = node;
        tail = node;
    } else {
        node->next = head;
        head->prev = node;
        head = node;
    }
}

float DoublyLinkedList::popHead(){
    if (head == nullptr){
        std::cerr << "popHead: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    }
    Node* old_head = head;
    if (head->next == nullptr) {
        head = nullptr;
        tail = nullptr;
    } else {
        head = head->next;
        head->prev = nullptr;
    }

    float value = *(old_head->value);
    delete old_head;

    return value;
}

float DoublyLinkedList::getHead(){
    if (head == nullptr){
        std::cerr << "getHead: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    } else {
        float value = *(head->value);
        return value;
    }
}

float DoublyLinkedList::getTail(){
    if (tail == nullptr){
        std::cerr << "getTail: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    } else {
        float value = *(tail->value);
        return value;
    }
}

int DoublyLinkedList::getLength(){
    Node* current = head;
    int n = 0;
    while (current != nullptr){
        current = current->next;
        n++;
    }
    return n;
}

void DoublyLinkedList::pushTail(float value){
    Node* node = new Node(value);
    if (tail == nullptr){
        head = node;
        tail = node;
    } else {
        tail->next = node;
        node->prev = tail;
        tail = node;
    }
}

// Debugging method used to ensure all necessary links between Nodes exist
bool DoublyLinkedList::is_connected(){
    if(head == nullptr){
        return true;
    }
    Node* current = head;
    while(current->next != nullptr){
        current = current->next;
    }
    if(current != tail){
        return false;
    }
    while(current->prev != nullptr){
        current = current->prev;
    }
    if(current != head){
        return false;
    }
    return true;
}

float DoublyLinkedList::popTail(){
    if (tail == nullptr){
        std::cerr << "popTail: List is empty" << std::endl;
        exit(EXIT_FAILURE);
    }
    float value = *(tail->value);
    if (tail->prev == nullptr) {
        delete(tail);
        tail = nullptr;
        head = nullptr;
    } else {
        Node* old_tail = tail;
        tail = tail->prev;
        tail->next = nullptr;
        delete(old_tail);
    }
    return value;
}

std::vector<float> DoublyLinkedList::getVector(){
    std::vector<float> vec;
    if (head == nullptr){
        return vec;
    } else {
        Node* current = head;
        while (current->next != nullptr) {
            vec.push_back(*(current->value));
            current = current->next;
        }
        vec.push_back(*(current->value));
        return vec;
    }
}

float DoublyLinkedList::valueAt(int index){
    int length = this->boundCheck(index);
    Node* current = head;
    for(int i = 0; i < index; i++){
        current = current->next;
    }
    float value = *(current->value);
    return value;
}

int DoublyLinkedList::boundCheck(int index){
    int length = this->getLength();
    if (index > length - 1 or index < 0) {
        std::cerr << "boundCheck: Index " << index << " out of range for list of length " << length << std::endl;
        exit(EXIT_FAILURE);
    }
    return length;
}

void DoublyLinkedList::deleteAt(int index){
    int length = this->boundCheck(index);
    if ((index == 0) and (head != tail)){
        Node* new_head = head->next;
        delete head;
        head = new_head;
        head->prev = nullptr;
    } else if ((index == 0) and (head == tail)) {
        delete head;
        head = nullptr;
        tail = nullptr;
    } else if ((index == length - 1) and (head != tail)) {
        Node* new_tail = tail->prev;
        delete tail;
        new_tail->next = nullptr;
        tail = new_tail;
    } else {
        Node* current = head;
        for(int i = 0; i < index-1; i++){
            current = current->next;
        }
        Node* node_to_delete = current->next;
        Node* node_after_deletion = node_to_delete->next;
        current->next = node_after_deletion;
        node_after_deletion->prev = current;
        delete node_to_delete;
    }
}

void DoublyLinkedList::insertAt(int index, float value){
    int length = this->boundCheck(index);
    Node* new_node = new Node(value);
    if (index == 0){
        new_node->next = head;
        head->prev = new_node;
        head = new_node;
    } else {
        Node* current = head;
        for(int i = 0; i < index-1; i++){
            current = current->next;
        }
        Node* node_after = current->next;
        current->next = new_node;
        new_node->next = node_after;
        new_node->prev = current;
        if (node_after != nullptr){
            node_after->prev = new_node;
        }
    } 
}

void DoublyLinkedList::reverse(){
    int length = this->getLength();
    if (length == 0 or length == 1){
        return;
    }
    Node* node1 = head;
    Node* node2 = head->next;
    head->next = nullptr;
    head->prev = node2;
    tail = node1;
    Node* node3;
    while(node2->next != nullptr){
        node3 = node2->next;
        node2->next = node1;
        node2->prev = node3;
        node1 = node2;
        node2 = node3;
    }
    node2->next = node1;
    node2->prev = nullptr;
    head = node2;
}

File: python/linkedLists/linkedLists.py
class SinglyLinkedList:
    class _Node:
        def __init__(self, item):
            self.item = item
            self.next = None
    
    def __init__(self, init_array = None):
        self.head = None
        if init_array:
            for item in reversed(init_array):
                self.push_head(item)

    def push_head(self,item):
        node = self._Node(item)
        if self.head is None:
            self.head = node
        else:
            node.next = self.head
            self.head = node

    def push_tail(self,item):
        tail = self._get_tail_node()
        node = self._Node(item)
        if tail is None:
            self.head = node
        else:
            tail.next = node

    def pop_head(self):
        if self.head is None:
            return None
        old_head = self.head
        if self.head.next is None:
            self.head = None
        else:
            self.head = self.head.next
        return old_head.item

    def pop_tail(self):
        if self.head is None:
            return None
        if self.head.next is None:
            old_head = self.head
            self.head = None
            return old_head.item
        length = self.get_length()
        current = self.head
        for _ in range(length-2):
            current = current.next
        new_tail = current
        old_tail = current.next
        new_tail.next = None
        return old_tail.item

    def get_length(self):
        node = self.head
        n = 0
        while node is not None:
            node = node.next
            n += 1
        return n

    def _get_tail_node(self):
        if self.head is None:
            return None
        current = self.head
        while current.next is not None:
            current = current.next
        return current
    
    def _get_node_at(self, index):
        # Does not do "in bounds" checks. Do these in the calling function.
        current = self.head
        for _ in range(index):
            current = current.next
        return current
    
    def get_head(self):
        if self.head is None:
            return None
        else:
            return self.head.item
    
    def get_tail(self):
        if self.head is None:
            return None
        else:
            return self._get_tail_node().item
    
    def item_at(self,index):
        length = self.get_length()
        if index >= length or index < 0:
            raise IndexError(f"Index of {index} out of range in list of length {length}")
        current = self.head
        for _ in range(index):
            current = current.next
        return current.item
    
    def get_list(self):
        l = []
        current = self.head
        while current is not None:
            l.append(current.item)
            current = current.next
        return l

    def delete_at(self, index):
        length = self.get_length()
        if index >= length or index < 0:
            raise IndexError(f"Index of {index} out of range in list of length {length}")
        if index == 0:
            old_head = self.head
            self.head = self.head.next
            del old_head
        else:
            node_before = self._get_node_at(index - 1)
            node_to_del = self._get_node_at(index)
            if index == length - 1:
                node_after = None
            else:
                node_after = self._get_node_at(index + 1)
            node_before.next = node_after
            del node_to_del

    def insert_at(self, item, index):
        # Inserts an item at specific index. The item at that index is pushed towards the tail (its index increases by one).
        length = self.get_length()
        if index >= length or index < 0:
            raise IndexError(f"Index of {index} out of range in list of length {length}")
        node = self._Node(item)
        if index == 0:
            node.next = self.head
            self.head = node
        else:
            node_before = self._get_node_at(index - 1)
            node_at_index = self._get_node_at(index)
            node_before.next = node
            node.next = node_at_index

    def reverse(self):
        if self.get_length() in [0,1]:
            return
        node1 = self.head
        node2 = self.head.next
        self.head.next = None
        while node2.next is not None:
            node3 = node2.next
            node2.next = node1
            node1 = node2
            node2 = node3
        node2.next = node1
        self.head = node2

class DoublyLinkedList:
    class _Node:
        def __init__(self, item):
            self.item = item
            self.next = None
            self.prev = None
    
    def __init__(self, init_array = None):
        self.head = None
        self.tail = None
        if init_array:
            for item in reversed(init_array):
                self.push_head(item)
    
    def push_head(self,item):
        node = self._Node(item)
        if self.head is None:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head.prev = node
            self.head = node
        
    def push_tail(self,item):
        node = self._Node(item)
        if self.tail is None:
            self.tail = node
            self.head = node
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node

    def pop_head(self):
        if self.head is None:
            return None
        old_head = self.head
        if self.head.next is None:
            self.head = None
            self.tail = None
        else:
            self.head.next.prev = None
            self.head = self.head.next
        return old_head
    
    def pop_tail(self):
        if self.tail is None:
            return None
        old_tail = self.tail
        if self.tail.prev is None:
            self.tail = None
            self.head = None
        else:
            self.tail.prev.next = None
            self.tail = self.tail.prev
        return old_tail
    
    def get_head(self):
        if self.head is None:
            return None
        else:
            return self.head.item
    
    def get_tail(self):
        if self.tail is None:
            return None
        else:
            return self.tail.item
    
    def get_length(self):
        node = self.head
        n = 0
        while node is not None:
            node = node.next
            n += 1
        return n
    
    def get_list(self):
        l = []
        current = self.head
        while current is not None:
            l.append(current.item)
            current = current.next
        return l
    
    def item_at(self,index):
        length = self.get_length()
        if index >= length or index < 0:
            raise IndexError(f"Index of {index} out of range in list of length {length}")
        current = self.head
        for _ in range(index):
            current = current.next
        return current.item
    
    def _get_node_at(self, index):
        # Does not do "in bounds" checks. Do these in the calling function.
        current = self.head
        for _ in range(index):
            current = current.next
        return current
    
    def delete_at(self, index):
        length = self.get_length()
        if index >= length or index < 0:
            raise IndexError(f"Index of {index} out of range in list of length {length}")
        if length == 1 and index == 0:
            single_item = self.head
            self.head = None
            self.tail = None
            del single_item
        elif index == 0:
            old_head = self.head
            self.head = self.head.next
            self.head.prev = None
            del old_head
        elif index == (length - 1):
            old_tail = self.tail
            self.tail = self.tail.prev
            self.tail.next = None
            del old_tail
        else:
            node_before = self._get_node_at(index - 1)
            node_to_del = self._get_node_at(index)
            node_after = self._get_node_at(index + 1)
            node_before.next = node_after
            node_after.prev = node_before
            del node_to_del

    def insert_at(self, item, index):
        # Inserts an item at specific index. The item at that index is pushed towards the tail (its index increases by one).
        length = self.get_length()
        if index >= length or index < 0:
            raise IndexError(f"Index of {index} out of range in list of length {length}")
        node = self._Node(item)
        if length == 1 and index == 0:
            item_at_index = self.head
            self.head = node
            self.head.next = item_at_index
            self.tail = item_at_index
            self.tail.prev = node
        elif index == 0:
            old_head = self.head
            self.head = node
            self.head.next = old_head
            old_head.prev = self.head
        else:
            node_before = self._get_node_at(index - 1)
            node_at_index = self._get_node_at(index)
            node_before.next = node
            node.next = node_at_index
            node.prev = node_before
            node_at_index.prev = node
    
    def reverse(self):
        if self.get_length() in [0,1]:
            return
        node1 = self.head
        self.tail = node1
        node2 = self.head.next
        self.head.next = None
        self.head.prev = node2
        while node2.next is not None:
            node3 = node2.next
            node2.next = node1
            node2.prev = node3
            node1 = node2
            node2 = node3
        node2.next = node1
        node2.prev = None
        self.head = node2
    
    def _is_all_connected(self):
        # Debugging function that checks all necessary links exits within the list
        if self.head is None and self.tail is None:
            return
        current = self.head
        while current.next is not None:
            current = current.next
        assert(current is self.tail)
        while current.prev is not None:
            current = current.prev
        assert(current is self.head)
        